### PR TITLE
 Countdown, data renewal and latest UI adjustment by spec

### PR DESCRIPTION
**Pull request by arealclimber**
**更新價格時間為15秒**
- 🐛 價格不一致 #263 

### DEVELOP

- [x] 132	PositionOpenModal
- [x] 133	PositionClosedModal
- [x] 134	PositionUpdatedModal
- [x] 114	TradeVisitorTab
- [x] 115	PositionVisitorTab
- [x] 050	OpenPositionItem
- [x] GlobalContext
- [x] MarketContext
- [x] 054	HistoryPositionModal
- [x] CryptoCard 
- [x] SignatureProcessModal 
- [x] TradeTab 
- [x] TradingHeader
- [x] PositionDetailsModal 
- [x] TickerSelectorBox 

### CHECK LIST

- [Naming convention](https://github.com/CAFECA-IO/WorkGuidelines/blob/main/technology/coding-convention/naming-convention.md) verification: checked
- Coding style verification: checked
- new Library: 0
- new Class / Component: 2
src/pages/404.tsx
src/pages/500.tsx
- new loop: 0
- new recursive: 0
- high risk: 0
- new sql: 0

### UML

- [Sequence Diagram 001 (interaction between user, frontend and backend)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00001.md)
- [Sequence Diagram 002 (background mechanism)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00002.md)
- [Component Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCP00001.md)
- [Class Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCD00001.md)
- [Global context document](https://www.notion.so/Global-Context-public-d93ac7c941c94fb6b173d34a7bc4e959)
- [Table of interface](https://github.com/CAFECA-IO/EventReport/blob/main/meeting/20230106_tbd_mvp.md)

### CODE REVIEW

File: src/contexts/market_context.tsx [hunk 1]
```diff
@@ -182,7 +182,7 @@ export const PositionInfoOnChart: ApexOptions = {
   //   dashArray: 2,
   // },
   annotations: {
-    position: 'back',
+    // position: 'back',
```
Comment (arealclimber) [inline src/contexts/market_context.tsx:185]: 刪掉 `PositionInfoOnChart`

File: src/components/position_closed_modal/position_closed_modal.tsx [hunk 2]
```diff
@@ -1,72 +1,310 @@
...
+    }
+    // async () => {
+    //   if (secondsLeft === 0) {
+    //     await wait(500);
+    //     setSecondsLeft(15);
+    //   }
+    // };
+
+    const intervalId = setInterval(() => {
+      setSecondsLeft(prevSeconds => prevSeconds - 1);
```
Comment (arealclimber) [inline src/components/position_closed_modal/position_closed_modal.tsx:284]: - 改成用終點線-打開視窗的時間戳

File: src/components/position_closed_modal/position_closed_modal.tsx [hunk 3]
```diff
@@ -1,72 +1,310 @@
...
+          : '',
+    });
+  };
+
+  useEffect(() => {
+    // if (!lock()) return;
+    setMounted(true);
+
+    if (!mounted) {
+      setPnlDisplayedStyle({
```
Comment (arealclimber) [inline src/components/position_closed_modal/position_closed_modal.tsx:244]: - PnL 顏色跟文字沒有正確顯示的問題

File: src/components/position_details_modal/position_details_modal.tsx [hunk 4]
```diff
@@ -256,6 +257,11 @@ const PositionDetailsModal = ({
       //   globalContext.visiblePositionDetailsModalHandler(false);
       //   // console.log('modal visible: ', modalVisible);
       // }, 1000);
+      globalCtx.dataPositionUpdatedModalHandler({
+        openCfdDetails: {...openCfdDetails},
+        updatedProps: {...changedProperties},
+      });
+      globalCtx.visiblePositionUpdatedModalHandler();
```
Comment (arealclimber) [inline src/components/position_details_modal/position_details_modal.tsx:264]: - input 及 toggle 跟 CFD 資訊不一致

Comment (arealclimber) [reply src/components/position_details_modal/position_details_modal.tsx:264]: #271

File: src/components/position_closed_modal/position_closed_modal.tsx [hunk 5]
```diff
@@ -1,72 +1,355 @@
...
+          value: randomIntFromInterval(0, 1000),
+        },
+      },
+      latestProps: {
+        // renewalDeadline: Date.now() / 1000 + POSITION_PRICE_RENEWAL_INTERVAL_SECONDS,
+        renewalDeadline: newTimestamp,
+        latestClosedPrice:
+          openCfdDetails.typeOfPosition === TypeOfPosition.BUY
+            ? randomIntFromInterval(
+                marketCtx.tickerLiveStatistics!.buyEstimatedFilledPrice * 0.75,
```
Comment (arealclimber) [inline src/components/position_closed_modal/position_closed_modal.tsx:197]: - 多單要關倉應該拿賣價
- 空單要關倉應該拿買價

File: src/components/position_closed_modal/position_closed_modal.tsx [hunk 6]
```diff
@@ -1,72 +1,355 @@
...
+    const newTimestamp = new Date().getTime() / 1000 + POSITION_PRICE_RENEWAL_INTERVAL_SECONDS;
+    setSecondsLeft(newTimestamp - Date.now() / 1000);
+
+    globalCtx.dataPositionClosedModalHandler({
+      openCfdDetails: {
+        ...openCfdDetails,
+        pnl: {
+          type: randomIntFromInterval(0, 100) <= 50 ? ProfitState.PROFIT : ProfitState.LOSS,
+          value: randomIntFromInterval(0, 1000),
+        },
```
Comment (arealclimber) [inline src/components/position_closed_modal/position_closed_modal.tsx:189]: TODO: 跟 userCtx 拿 getOpenCFD(id)

File: src/components/position_open_modal/position_open_modal.tsx [hunk 7]
```diff
@@ -1,68 +1,279 @@
...
+        price:
+          openCfdRequest.typeOfPosition === TypeOfPosition.BUY
+            ? randomIntFromInterval(
+                marketCtx.tickerLiveStatistics!.buyEstimatedFilledPrice * 0.75,
+                marketCtx.tickerLiveStatistics!.buyEstimatedFilledPrice * 1.25
+              )
+            : openCfdRequest.typeOfPosition === TypeOfPosition.SELL
+            ? randomIntFromInterval(
+                marketCtx.tickerLiveStatistics!.sellEstimatedFilledPrice * 1.1,
+                marketCtx.tickerLiveStatistics!.sellEstimatedFilledPrice * 1.25
```
Comment (arealclimber) [inline src/components/position_open_modal/position_open_modal.tsx:178]: - 多單要關倉應該拿賣價
- 空單要關倉應該拿買價

File: src/components/position_closed_modal/position_closed_modal.tsx (hunk 6 above)
Comment (arealclimber) [reply src/components/position_closed_modal/position_closed_modal.tsx:190]: https://github.com/CAFECA-IO/TideBit-DeFi/issues/287

File: src/components/position_open_modal/position_open_modal.tsx [hunk 8]
```diff
@@ -1,68 +1,279 @@
...
+  const submitClickHandler = async () => {
+    const [lock, unlock] = locker('position_open_modal.submitClickHandler');
 
-  // TODO: update order function
-  const submitClickHandler = () => {
+    if (!lock()) return;
+    await wait(DELAYED_HIDDEN_SECONDS / 5);
     modalClickHandler();
+
+    globalCtx.dataLoadingModalHandler({
```
Comment (arealclimber) [inline src/components/position_open_modal/position_open_modal.tsx:60]: i18n

Comment (arealclimber) [reply src/components/position_open_modal/position_open_modal.tsx:60]: https://github.com/CAFECA-IO/TideBit-DeFi/issues/287

**Review by Luphia — APPROVED** [review]
- no bug found

**Pull request merged**